indico/modules/events/timetable/client/js/TimetableCreateModal.tsx

```diff
@@ -0,0 +1,343 @@
+
+  const personLinkFieldParams = {
+    allowAuthors: true,
+    canEnterManually: true,
+    defaultSearchExternal: false,
+    extraParams: {},
+    hasPredefinedAffiliations: true,
+    nameFormat: 'first_last',
+  };
+
+  const initialValues: DraftEntry = {
+    title: '',
+    description: '',
+    person_links: [],
+    keywords: [],
+    references: [],
+    location_data: {inheriting: false},
+    conveners: [],
+    start_dt: entry.startDt.format('YYYY-MM-DDTHH:mm:ss'),
+    duration: entry.duration * 60, // Minutes to seconds
+    session_id: null,
+    code: null,
+  };
+
+  const sessions = useSelector(selectors.getSessions);
+  // const eventStartDt = useSelector(selectors.getEventStartDt);
+  // const eventEndDt = useSelector(selectors.getEventEndDt);
+
+  const sessionValues: Session[] = Object.values(sessions);
+
+  const extraOptions = {
+    minStartDt: useSelector(selectors.getEventStartDt),
+    maxEndDt: useSelector(selectors.getEventEndDt),
+  };
+
+  const forms: {[key in EntryType]: React.ReactElement} = {
+    [EntryType.Contribution]: (
+      <ContributionFormFields
+        eventId={eventId}
+        initialValues={initialValues}
+        personLinkFieldParams={personLinkFieldParams}
+        extraOptions={extraOptions}
+      />
+    ),
+    [EntryType.SessionBlock]: sessionValues.length ? (
+      <>
+        <SessionSelect sessions={sessionValues} required />
+        <SessionBlockFormFields eventId={eventId} extraOptions={extraOptions} />
+      </>
+    ) : (
+      <Message
+        icon="question circle"
+        header={Translate.string('No sessions available')}
+        color="yellow"
+        content={Translate.string('Please create a session before creating a session block.')}
+      />
+    ),
+    [EntryType.Break]: (
+      <BreakFormFields
+        eventId={eventId}
+        locationParent={undefined}
+        initialValues={initialValues}
+        extraOptions={extraOptions}
+      />
+    ),
+  };
+
+  // TODO: (Ajob) Implement properly in next issue on editing existing entries
+  const [activeForm, setActiveForm] = useState(isEditing ? entry['type'] : Object.keys(forms)[0]);
+
+  const _mapPersonLinkToSchema = data => {
+    return {
+      title: data.title,
+      name: data.name,
+      first_name: data.firstName,
+      last_name: data.lastName,
+      affiliation: data.affiliation,
+      affiliation_id: data.affiliationId,
+      email: data.email,
+      address: data.address,
+      phone: data.phone,
+      roles: data.roles,
+      type: data.type,
+      avatar_url: data.avatarURL,
+    };
+  };
+
+  const _mapDataToEntry = (data): TopLevelEntry => {
+    const {object} = data;
+    delete data['object'];
+    data['duration'] /= 60; // Seconds to minutes
+    const {
+      id,
+      title,
+      type: objType,
+      duration,
+      start_dt: startDt,
+      colors,
+      y,
+      x,
+      column,
+      maxColumn,
+      session_id: sessionId,
+    } = {...object, ...data};
```

Comment:
This feels like future maintenance hell... are the values always expected to be in one of the two places? If yes, I'd rather destructure them separately.

Or even better, instead of mutating data above (that feels quite dangerous anyway), why not just do it like this?

```js
const {
  object: {id, title, ...objectRest},
  whatever,
  you,
  have,
  here,
  ...rest
} = data;
```

(skip the rest stuff if you don't actually need it)

Reply:
You're right. Fixed